### ISSUE TITLE
3.9.4: chunkserver errors "EBADF (Bad file descriptor)"

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#349]
On start of chunkservers, during scan of assets the following errors are logged for various chunks:

~~~~
write_block_to_chunk: file:/var/lib/lizardfs/34/chunks99/chunk_0000000006994176_00000001.mfs - write error: EBADF (Bad file descriptor)
chunk_writecrc: file:/var/lib/lizardfs/34/chunks99/chunk_0000000006994176_00000001.mfs - write error: EBADF (Bad file descriptor)
hdd_io_end: file:/var/lib/lizardfs/34/chunks99/chunk_0000000006994176_00000001.mfs - write error: EBADF (Bad file descriptor)
chunk_writecrc: file:/var/lib/lizardfs/34/chunks99/chunk_0000000006994176_00000001.mfs - write error: EBADF (Bad file descriptor)
hdd_io_end: file:/var/lib/lizardfs/34/chunks99/chunk_0000000006994176_00000001.mfs - write error: EBADF (Bad file descriptor)
replication error: failed to close chunk (IO error)
~~~~

The following error is repeated 50 times:
~~~~
write_block_to_chunk: file:/var/lib/lizardfs/34/chunksA2/chunk_0000000001A2FD2C_00000002.liz - crc write error: EBADF (Bad file descriptor)
~~~~

As far as I can tell those chunk files do not exist.

**Comment by onlyjob** [comment]
Also `hdd_io_end: refcount = 0 - This should never happen!` appear in chunkservers' logs fairly often during scanning...

**Comment by onlyjob** [comment]
Any update regarding this issue please?
Presence of "Bad file descriptor" errors pollute "last error" field in CGI/Disks.
Also this error makes it hard(er) to feel confidence in LizardFS...

**Comment by onlyjob** [comment]
This happens even on brand new chunkserver without legacy chunks (i.e. configured with `CREATE_NEW_CHUNKS_IN_MOOSEFS_FORMAT = 0`):

~~~~
write_block_to_chunk: file:/var/lib/lizardfs/01/chunksF4/chunk_0000000007F48E34_00000001.liz - crc write error: EBADF (Bad file descriptor)
write_block_to_chunk: file:/var/lib/lizardfs/01/chunksF4/chunk_0000000007F48E34_00000001.liz - crc write error: EBADF (Bad file descriptor)
hdd_io_end: refcount = 0 - This should never happen!
~~~~

**Comment by psarna** [comment]
I remember that you changed values of
~~~
#define DELAYEDSTEP 2
#define OPENDELAY 5
~~~
in code(*chunkserver/hddspacemgr.cc*) when we were troubleshooting https://github.com/lizardfs/lizardfs/issues/330,
are there still changed to lower values? Maybe some descriptors are prematurely closed due to this.

**Comment by onlyjob** [comment]
> Maybe some descriptors are prematurely closed

During scan? I doubt that as errors appear only during scan after start of chunkserver...

Anyway I reverted `DELAYEDSTEP` and `OPENDELAY` to default values but it has no effect on the problem.

**Comment by psarna** [comment]
Do you have any unconventional configuration - very short REPLICATION_DELAY_DISCONNECT or something similar? Are there any other logs that seem to corelate (perhaps in master, mount)?

I've never seen a single EBADF error in any of our 3.9.4 installations and I can't reproduce it.

**Comment by onlyjob** [comment]
On master side there are few `replication status: Chunk already exists` errors that are most likely related to #350. `REPLICATIONS_DELAY_DISCONNECT` is set to `172800`.

> I've never seen a single EBADF error in any of our 3.9.4 installations and I can't reproduce it.

I see many of those very annoying errors on most chunkservers during their initialisation.
Errors more likely to appear on chunkservers with multiple HDDs and on HDDs with many chunks. Sometimes those errors appear on any HDD including ones marked for removal.

There are no errors from client and I can't think of anything particular in my configuration that can be responsible for this. Given recent history of data loss and descriptor-related problems those errors make me feel very uncomfortable and insecure...

**Comment by onlyjob** [comment]
Problem is probably introduced with 89e339a5 (see [comment](https://github.com/lizardfs/lizardfs/commit/89e339a50fa10566e4c0309d49006c2c26c700c2#commitcomment-14588507)) and/or e61487d.

**Comment by onlyjob** [comment]
Reproducible: always, by simple restart of one chunkserver when all chunkservers  are up and running. Also reproducible by restarting chunkserver when some chunkservers are down.
Errors often manifest on more than one HDD. Errors randomly distributed but more likely to  affect HDDs with greater number of chunks.

**Comment by onlyjob** [comment]
Not sure if it is related but I found many `chunk: 00000000017FE14C replication status: Unknown LizardFS error` in master's log for various chunks... Please investigate.

**Comment by psarna** [comment]
Nope, it may happen after b131de1caa5fe582c90009eeb1f92e40ec0c6bf0 and just
means, that description for this error wasn't assigned yet. The important
fact is that if you see this, it means that master properly handled
something like replication timeout or connection hiccup. It's totally
harmless and once proper codes are assigned to each situation you won't see
"unknown error" logs anymore.

**Comment by onlyjob** [comment]
`write_block_to_chunk` errors sometimes are followed by `serious error: crc changes lost` -- all due to `write error: EBADF (Bad file descriptor)` errors. There are so many of them... Any ideas?

**Comment by onlyjob** [comment]
Errors only appear during scan and apparently none of chunk files logged with `EBADF` exist...

**Comment by onlyjob** [comment]
More details: I've made a list of all chunks before starting a particular chunkserver:

    find /var/lib/lizardfs -type f -name "chunk_*" -print > allchunks.txt

There are 2_779_921 chunks on chunkserver. Open files limit is 32768.
On chunkserver start the following error was logged (among many other similar errors):

~~~~
mfschunkserver[18933]: write_block_to_chunk: file:/var/lib/lizardfs/01/chunksBF/chunk_0000000009BFECEA_00000001.liz - crc write error: EBADF (Bad file d
mfschunkserver[18933]: hdd_io_end: file:/var/lib/lizardfs/01/chunksBF/chunk_0000000009BFECEA_00000001.liz - fsync (direct call) error: EBADF (Bad file d
mfschunkserver[18933]: hdd_io_end: file:/var/lib/lizardfs/01/chunksBF/chunk_0000000009BFECEA_00000001.liz - fsync (direct call) error: EBADF (Bad file d
mfschunkserver[18933]: replication error: failed to close chunk (IO error)
~~~~

At the same time master logged:

    mfsmaster[27538]: (192.168.0.2:9422) chunk: 0000000009BFECEA replication status: IO error

After chunkserver initialisation chunk `0000000009BFECEA` disappeared from file system.
Nearly all chunkservers affected by this problem (though I rarely see it on few lightweight chunkservers that finish scanning their HDDs fast).

This is one of the sloppiest bugs I've seen and possible manifestation of #350.
Would you guys give it some attention please?

**Comment by onlyjob** [comment]
And this is not a harmless bug as it causes deletion of valid chunks on chunkserver restart. :-(

**Comment by onlyjob** [comment]
Setting `HDD_LEAVE_SPACE_DEFAULT = 4444GiB` prevents replication to chunkserver and manifestation of  this error from happening therefore proving that the bug occur in replication during scan.

**Comment by onlyjob** [comment]
Interesting to note that another copy of chunk `0000000009BFECEA` is in _MFS_ format. While chunkserver holding _MFS_ version of this chunk is running, another chunkserver (configured with `CREATE_NEW_CHUNKS_IN_MOOSEFS_FORMAT=0`) is restarted. Before restart the latter had chunk `0000000009BFECEA` in _LIZ_ format but that chunk was deleted following `EBADF` error.

**Comment by psarna** [comment]
Oook, your latest insights were very helpful. I wondered why replication occurs while hdd scan, since you had REPLICATION_DELAY_DISCONNECT set to a pretty high value. It looks like it's an issue with rebalancing, which *could* be performed concurrently with folder scan, which is bad.

Please try
http://cr.skytechnology.pl:8081/#/c/2327/

**Comment by onlyjob** [comment]
Thank you very much, @psarna. 069139a fixed the problem. Now chunkservers start much faster too. :)

I'm surprised that it took so long to identify this bug. Heavy chunkservers can scan for ~10 minutes and replication used to start long before scan is finished slowing down join and causing errors. I'm very happy that this super-annoying bug is fixed.

**Comment by 4Dolio** [comment]
Yes  TY. My delay is one hours as some of our large jbods can take over 30 minutes to scan.. looking forward to much faster...

**Comment by onlyjob** [comment]
@4Dolio, true, some 3TB HDDs with millions of chunks can scan for 20+ minutes.

Apparently this problem is not completely fixed. I've just reproduced it again using the following steps:

  1 Start chunkserver with one HDD mount point not mounted. After chunkserver initialisation one disk is detected as "damaged".
  2 Mount HDD.
  3 Reload chunkserver. During scanning of HDD chunkserver logs the following (just as before):

~~~~
write_block_to_chunk: file:{chunkfilename}.liz - crc write error: EBADF (Bad file descriptor)
replication error: failed to write chunk (No such chunk)
~~~~

**Comment by onlyjob** [comment]
@blink69, why did you close this bug? It was re-opened with another case that needs fixing.

**Comment by blink69** [comment]
Sorry my bad, we will fix this
